server/src/main/java/io/druid/segment/loading/SegmentLoaderLocalCacheManager.java

```diff
@@ -104,53 +104,71 @@ public File getSegmentFiles(DataSegment segment) throws SegmentLoadingException
-        if (!storageDir.mkdirs()) {
-          log.debug("Unable to make parent file[%s]", storageDir);
-        }
+        // We use a marker to prevent the case where a segment is downloaded, but before the download completes,
+        // the parent directories of the segment are removed
+        final File downloadStartMarker = new File(storageDir, "downloadStartMarker");
         try {
-          if (!downloadStartMarker.createNewFile()) {
-            throw new SegmentLoadingException("Was not able to create new download marker for [%s]", storageDir);
+          synchronized (lock) {
+            if (!storageDir.mkdirs()) {
+              log.debug("Unable to make parent file[%s]", storageDir);
+            }
+            try {
+              if (!downloadStartMarker.createNewFile()) {
+                throw new SegmentLoadingException("Was not able to create new download marker for [%s]", storageDir);
+              }
+            }
+            catch (IOException e) {
+              throw new SegmentLoadingException(e, "Unable to create marker file for [%s]", storageDir);
+            }
           }
-        }
-        catch (IOException e) {
-          throw new SegmentLoadingException(e, "Unable to create marker file for [%s]", storageDir);
-        }
-      }
 
-      // LoadSpec isn't materialized until here so that any system can interpret Segment without having to have all the LoadSpec dependencies.
-      final LoadSpec loadSpec = jsonMapper.convertValue(segment.getLoadSpec(), LoadSpec.class);
-      final LoadSpec.LoadSpecResult result = loadSpec.loadSegment(storageDir);
-      if(result.getSize() != segment.getSize()){
-        log.warn("Segment [%s] is different than expected size. Expected [%d] found [%d]", segment.getIdentifier(), segment.getSize(), result.getSize());
-      }
+          // LoadSpec isn't materialized until here so that any system can interpret Segment without having to have all the LoadSpec dependencies.
+          final LoadSpec loadSpec = jsonMapper.convertValue(segment.getLoadSpec(), LoadSpec.class);
+          final LoadSpec.LoadSpecResult result = loadSpec.loadSegment(storageDir);
+          if(result.getSize() != segment.getSize()){
+            log.warn("Segment [%s] is different than expected size. Expected [%d] found [%d]", segment.getIdentifier(), segment.getSize(), result.getSize());
+          }
 
-      if (!downloadStartMarker.delete()) {
-        throw new SegmentLoadingException("Unable to remove marker file for [%s]", storageDir);
-      }
+          if (!downloadStartMarker.delete()) {
+            throw new SegmentLoadingException("Unable to remove marker file for [%s]", storageDir);
+          }
+
+          retVal = storageDir;
+          break;
+        }
+        catch (Exception e) {
+          if (!locIter.hasNext()) {
+            throw e;
+          }
+          log.makeAlert(e, "Failed to load segment in current location %s, try next location", loc.getPath().getAbsolutePath())
```

Comment:
Isn't it better to alert the administrator there is a disk failure? Otherwise, the administrator may never know about that there is a disk should be replace.